### PR TITLE
[docs] new attention mask helpers

### PR DESCRIPTION
**Pull request by stevhliu**
adds docs for #43924 on using new `create_*_mask` functions

### CODE REVIEW

File: docs/source/en/attention_interface.md [hunk 1]
```diff
@@ -182,9 +182,94 @@ model(torch.ones(1, 5, dtype=int), a_new_kwargs=..., another_new_kwargs=...)
...
+    encoder_hidden_states=encoder_hidden_states,
+)
+```
+
+</hfoption>
+</hfoptions>
+
+Add extra constraints on top of the base mask with the `or_mask_function` and `and_mask_function` arguments. Use `or_mask_function` to let additional positions attend, and `and_mask_function` to restrict the base pattern further. Both follow the 4-index `mask_function` signature described in [AttentionMaskInterface](#attentionmaskinterface), they take `(batch_idx, head_idx, q_idx, kv_idx)` and return a boolean.
+
+For example, Gemma 3 uses `or_mask_function` to flip a causal mask into a fully bidirectional one when `config.use_bidirectional_attention` is enabled. The overlay returns `True` for every position, so the union with the causal pattern lets every token attend to every other token.
```
Comment (vasqu) [inline docs/source/en/attention_interface.md]: We should have a warning here that these and/or masks are powerful to express any pattern but in turn are slower - this can be noticable on slower models especially (e.g. 200M) - and are not compatible with executorch

File: docs/source/en/attention_interface.md [hunk 2]
```diff
@@ -182,9 +182,94 @@ model(torch.ones(1, 5, dtype=int), a_new_kwargs=..., another_new_kwargs=...)
...
+    "config": self.config,
+    "inputs_embeds": inputs_embeds,
+    "attention_mask": attention_mask,
+    "past_key_values": past_key_values,
+    "position_ids": position_ids,
+}
+if self.config.use_bidirectional_attention:
+    mask_kwargs["or_mask_function"] = lambda *args: torch.tensor(True, dtype=torch.bool)
+
+attention_mask = create_causal_mask(**mask_kwargs)
```
Comment (vasqu) [inline docs/source/en/attention_interface.md:305]: I would not reference gemma here but just keep it as general example --> how to turn causal to bidirectional etc

I need to refactor this because it's not so efficient 😬

File: docs/source/en/attention_interface.md [hunk 3]
```diff
@@ -182,9 +182,94 @@ model(torch.ones(1, 5, dtype=int), a_new_kwargs=..., another_new_kwargs=...)
...
+```
+
+During generation, [`~GenerationMixin.generate`] builds masks through [`create_masks_for_generate`], which dispatches to the right `create_*_mask` based on the model config. Override it on a model class to plug in a custom masking strategy for generation.
+
+## AttentionMaskInterface
+
+[`AttentionMaskInterface`] is the registry the `create_*_mask` functions consult to convert a mask into the format the active attention backend expects. FlexAttention needs a [BlockMask](https://docs.pytorch.org/docs/stable/nn.attention.flex_attention.html#torch.nn.attention.flex_attention.BlockMask), SDPA needs a 4D tensor, and FlashAttention doesn't need a mask. Register a custom backend, or override the formatter for an existing one, with [`AttentionMaskInterface.register`].
```
Comment (vasqu) [inline docs/source/en/attention_interface.md]: not sure but maybe we could reorder to have the same order as in the base attention interface -> how to build?

Comment (vasqu) [reply docs/source/en/attention_interface.md]: FA needs a base 2D padding mask

**Review by vasqu — APPROVED** [review]
Just some smaller comments. I'd be careful about promoting the and/or masks as they are expensive and not executorch compatible (so users should be aware imo)
